feat: 향수 검색 로직 변경

xyzzeong opened the pull request:
- 검색 타입 필터 추가
    - 전체 검색 시 브랜드 1개 + 향수 30개씩
    - 브랜드 검색 시 브랜드만 30개씩
    - 향수 검색 시 향수만 30개씩
- 페이징 처리

* PerfumeSimpleVo에 브랜드 네임 추가 작업은 conflit 방지로 제거하였음

xyzzeong committed:
feat: 향수 검색 로직 변경
- 검색 타입 필터 추가
    - 전체 검색 시 브랜드 1개 + 향수 30개씩
    - 브랜드 검색 시 브랜드만 30개씩
    - 향수 검색 시 향수만 30개씩
- 페이징 처리

xyzzeong committed:
Merge branch 'develop' into feature/129_search_paging

junhaesung commented on acm-api/src/main/kotlin/mashup/backend/spring/acm/application/search/SearchApplicationService.kt:
```diff
@@ -2,17 +2,35 @@ package mashup.backend.spring.acm.application.search
+        perfumes = perfumeService.searchByName(name = name, pageable = pageable)
+    )
+
+    fun searchWithBrand(name: String, pageable: Pageable): SearchResponseVo = SearchResponseVo(
```
메서드 네이밍이 기능과 일치하지 않아서 오해할 수 있을거같아요
직역하면 '브랜드로 검색' 인데, 하는 일은 '이름으로 브랜드만 검색' 이어서요.

xyzzeong replied:
네! 그럼 `이름으로 브랜드만 검색` -> `search only brands by name`인데,
좀 긴 것 같아서 `searchBrandsByName`으로 변경하겠습니다~!

junhaesung commented on acm-api/src/main/kotlin/mashup/backend/spring/acm/application/search/SearchApplicationService.kt:
```diff
@@ -2,17 +2,35 @@ package mashup.backend.spring.acm.application.search
+        perfumes = emptyList()
+    )
+
+    fun searchWithPerfume(name: String, pageable: Pageable): SearchResponseVo = SearchResponseVo(
```
바로 위 내용과 동일하게 메서드 이름과 기능이 일치하면 좋을거같아요

xyzzeong replied:
이것도 마찬가지로 `searchPerfumesByName`으로 변경할게요~

junhaesung commented on acm-api/src/main/kotlin/mashup/backend/spring/acm/presentation/api/search/SearchData.kt:
```diff
@@ -2,14 +2,31 @@ package mashup.backend.spring.acm.presentation.api.search
+    @field:NotBlank
+    val type: String?,
+    @field:Min(0)
+    val page: Int? = 0,
```
기본값이 있는데 nullable 타입을 사용하신 이유가 있을까요? 
not null 타입으로 사용해도 좋을거같아서요

xyzzeong replied:
값이 안 들어올 것을 생각해서 nullable로 한거였는데 기본 값이 있었네요🤭
그러면 이와 동일하게 size도 not null로 적용하면 좋을까요?

junhaesung replied:
네 size 도 not null 적용하면 좋겠네요!

junhaesung commented on acm-api/src/main/kotlin/mashup/backend/spring/acm/presentation/api/search/SearchData.kt:
```diff
@@ -2,14 +2,31 @@ package mashup.backend.spring.acm.presentation.api.search
-)
\ No newline at end of file
+    @field:NotBlank
+    val type: String?,
```
요청하는 값을 enum 으로 변환하는 경우에는 enum 을 직접 사용하는것도 편리할거같아요. 
enum 으로 쓰면 swagger 에서 자동으로 model 에 어떤 원소들이 있는지 같이 나타내주더라고요
MvcConfig.addFormatters 에 enum converting 관련 설정 있는데 참고해도 좋을거같아요

junhaesung commented on acm-api/src/main/kotlin/mashup/backend/spring/acm/presentation/api/search/SearchData.kt:
```diff
@@ -2,14 +2,31 @@ package mashup.backend.spring.acm.presentation.api.search
 import javax.validation.constraints.NotBlank
 
-data class SearchData(
+data class SearchResponse(
```
ApiResponse 의 data 필드에 사용하는 값은 ~data 로 이름 맞추어도 좋을거같아요!
다른 API 응답 모델에서는 이런식으로 이름 짓고있어서요

junhaesung commented on acm-domain/src/main/kotlin/mashup/backend/spring/acm/domain/brand/BrandService.kt:
```diff
@@ -50,9 +51,12 @@ class BrandServiceImpl(
+    override fun searchByName(name: String, pageable: Pageable): List<BrandSimpleVo> = brandRepository.findByNameContaining(name, pageable)
         .map { BrandSimpleVo(it) }
 
+    override fun searchOneByName(name: String): List<BrandSimpleVo> = brandRepository.findFirstByNameContaining(name)
```
이 메서드는 54번 라인의 `searchByName(name: String, pageable: Pageable)` 에 포함되는 기능이어서, 기존 로직을 재사용해도 좋을거같아요. 
호출하는쪽에서 pageable 값만 바꾸면 될거같아서요. 

메서드를 분리하는게 더 좋다고 생각하시면 이대로 유지하는것도 좋을거같습니다

xyzzeong replied:
오 무슨 말씀이신지 이해했습니다!! 이 부분 고려해볼게요~

junhaesung reviewed (approved):
수고하셨습니다!
코멘트 몇 개 추가했어요-
로컬환경이나, 배포하고나서 개발서버에서 잘 동작하는지 테스트 한번씩 부탁드려요

xyzzeong committed:
refactor: 향수 검색 코드 리뷰 반영
- SearchRequest 기본값이 있는 프로퍼티는 not null로 수정
- SearchRequest에서 SearchType 프로퍼티 String? -> Enum 타입 수정
- SearchResponse -> SearchData
- searchOneByName 제거 및 기존 메소드 재사용

junhaesung commented:
👍
